src/key_funcs.c

```diff
@@ -74,9 +74,40 @@ void free_Array(struct Array *key_array) {
     free(key_array->array);
 }
 
+void start_tx(char **query,  size_t *current_len) {
+    char *begin = "BEGIN; ";
+    strcpy(*query, begin);
+    *current_len += strlen(begin);
+}
+
+void end_tx(char **query, size_t *current_len) {
+    char *commit = "COMMIT;";
+    strcat(*query + *current_len, commit);
+    *current_len += strlen(commit);
+    (*query)[*current_len] = '\0';
+}
+
+
+int resize_check(char *value, char **query, size_t *current_len, int *q_size) {
+    int val_len = strlen(value);
+    if (val_len + *current_len >= *q_size) {
+        *query = realloc(*query, 2 * (*q_size) * sizeof(char));
+        if (*query == NULL) {
+            perror("realloc");
+            return 1;
+        }
+        (*q_size) *= 2;
+    }
+    memcpy(*query + (*current_len), value, val_len + 1);
+    *current_len += val_len;
+    sqlite3_free(value);
+    return 0;
+}
+
 
 int prepare_query(struct Array *arr, char **query, int query_size, int type) {
     *query = malloc(sizeof(char) * query_size * arr->used);
+    query_size *= arr->used; // current size of the query
```

Comment:
Additional note, while working on this card I noticed that `realloc()` was being called a lot. After a quick investigation, I realized query_size was not set to the actual size in bytes of the query, but rather the size of a single query.

For example, one query would be: SELECT * FROM keys WHERE seed LIKE 'hello';
The query string is actually 0+ queries, and so query_size should have been set to the number of bytes of all queries combined.